src/halmos/__main__.py

```diff
@@ -394,7 +396,193 @@ def is_global_fail_set(context: CallContext) -> bool:
     return hevm_fail or any(is_global_fail_set(x) for x in context.subcalls())
 
 
-def run_test(ctx: FunctionContext) -> TestResult:
+# execute invariant test functions in multiple depths
+# reuse states at each depth for all invariant functions
+def run_invariant_tests(ctx, setup_ex):
+    # check all invariants against the setup state
+    test_results = run_tests(ctx, setup_ex, ctx.invariant_funsigs, terminal=False)
+
+    # accumulated test results; to be updated later
+    test_results_map = {r.name: r for r in test_results if r.exitcode != 0}
+
+    # invariant tests that have not failed yet
+    funsigs = [r.name for r in test_results if r.exitcode == 0]
+
+    # if no more invariant tests to run, stop
+    if not funsigs:
+        return test_results_map.values()
+
+    exs = [setup_ex]
+
+    depth = 0
+    while True:
+        depth += 1
+        exs, funsigs = run_single_invariant_step(
+            ctx, exs, funsigs, test_results_map, depth
+        )
+
+        # todo: merge, simplify, or prioritize exs to mitigate path explosion
```

Comment:
here we can add more optimization to reduce the number of paths

Reply:
some ideas:
- filter out infeasible paths with a longer branching timeout
- prioritize paths based on uniqueness --- e.g., uncovered paths vs variations in byte sizes
- merge similar paths that have the same storage but slightly different path constraints